common/flatpak-utils.c

```diff
@@ -6411,12 +6411,122 @@ flatpak_progress_new (FlatpakProgressCallback progress,
+  COPY_OSTREE_STATE ("outstanding-metadata-fetches");
+  COPY_OSTREE_STATE ("outstanding-writes");
+  COPY_OSTREE_STATE ("scanning");
+  COPY_OSTREE_STATE ("scanned-metadata");
+  COPY_OSTREE_STATE ("fetched-delta-parts");
+  COPY_OSTREE_STATE ("total-delta-parts");
+  COPY_OSTREE_STATE ("fetched-delta-fallbacks");
+  COPY_OSTREE_STATE ("total-delta-fallbacks");
+  COPY_OSTREE_STATE ("fetched-delta-part-size");
+  COPY_OSTREE_STATE ("total-delta-part-size");
+  COPY_OSTREE_STATE ("bytes-transferred");
+  COPY_OSTREE_STATE ("fetched");
+  COPY_OSTREE_STATE ("metadata-fetched");
+  COPY_OSTREE_STATE ("requested");
+  COPY_OSTREE_STATE ("start-time");
+  COPY_OSTREE_STATE ("status");
+  COPY_OSTREE_STATE ("outstanding-extra-data");
+  COPY_OSTREE_STATE ("total-extra-data-bytes");
+  COPY_OSTREE_STATE ("transferred-extra-data-bytes");
+  COPY_OSTREE_STATE ("downloading-extra-data");
+
+#undef COPY_OSTREE_STATE
+}
+
+static void
+handle_chained_progress (OstreeAsyncProgress *chained_progress,
+                         gpointer             user_data)
+{
+  OstreeAsyncProgress *original_progress = (OstreeAsyncProgress *) user_data;
+
+  /* Sync the chained progress's state back to the original instance, to take
+   * into account any updates received while a different GMainContext was
+   * active */
+  copy_ostree_progress_state (chained_progress, original_progress);
+
+  gpointer original_data = g_object_get_data (G_OBJECT (original_progress), "callback_data");
+  progress_cb (original_progress, original_data);
+}
+
+/*
+ * This is necessary when pushing a temporary GMainContext to be the thread
+ * default with flatpak_main_context_new_default() in order to call an async
+ * operation as if it were sync, if you have an OstreeAsyncProgress object that
+ * would otherwise be forwarded into the async operation.
+ *
+ * This is because the original OstreeAsyncProgress object won't receive any
+ * signals while the temporary GMainContext is active, since the GMainContext it
+ * was created with won't be iterated.
+ *
+ * tl;dr instead of this:
+ *
+ *     my_operation (OstreeAsyncProgress *progress)
+ *     {
+ *       g_autoptr(GMainContextPopDefault) context = NULL;
+ *       context = flatpak_main_context_new_default ();
+ *       ostree_some_async_op (progress, some_callback_setting_some_flag, data);
+ *       while (wait_for_flag)
+ *         g_main_context_iteration (context, TRUE);
+ *     }
+ *
+ * do this:
+ *
+ *     my_operation (OstreeAsyncProgress *progress)
+ *     {
+ *       g_autoptr(GMainContextPopDefault) context = NULL;
+ *       g_autoptr(OstreeAsyncProgressChain) chained_progress = NULL;
+ *       context = flatpak_main_context_new_default ();
+ *       chained_progress = flatpak_progress_chain (progress);
+ *       ostree_some_async_op (chained_progress,
+ *                             some_callback_setting_some_flag, data);
+ *       while (wait_for_flag)
+ *         g_main_context_iteration (context, TRUE);
+ *     }
```

Comment:
Might be worth mentioning in this comment that `OstreeAsyncProgressChained` is only safe to use when the two `GMainContext`s are in the same thread. Otherwise it will call the outer callback function from the wrong thread and cause thread safety issues.